File: leetcode/501/solution.go
// https://leetcode.com/problems/find-mode-in-binary-search-tree/
// 501. Find Mode in Binary Search Tree

package main

func findMode(root *TreeNode) []int {
	// q is a simple version of queue
	// (Using index i allows us to avoid unnecessary pop operation that costs O(n) complexity in Go)
	var q []*TreeNode
	i := 0
	if root != nil {
		q = append(q, root)
	}

	// populate frequency for all numbers using BFS
	freq := make(map[int]int)
	for i < len(q) {
		node := q[i]
		i += 1
		freq[node.Val]++
		if node.Left != nil {
			q = append(q, node.Left)
		}
		if node.Right != nil {
			q = append(q, node.Right)
		}
	}

	// find out the maximum frequency
	maxValue := 0
	for _, count := range freq {
		if count > maxValue {
			maxValue = count
		}
	}

	// collect all numbers that appears maxValue times
	var arr []int
	for n, count := range freq {
		if count == maxValue {
			arr = append(arr, n)
		}
	}

	return arr
}

File: leetcode/501/solution.rs
use std::cell::RefCell;
/**
 * https://leetcode.com/problems/find-mode-in-binary-search-tree/
 * 501. Find Mode in Binary Search Tree
 */
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
impl Solution {
    pub fn find_mode(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut freq = std::collections::HashMap::new();
        let mut queue = std::collections::VecDeque::new();

        if let Some(node) = root {
            queue.push_back(node);
        }

        while let Some(node) = queue.pop_front() {
            let node = node.borrow();
            *freq.entry(node.val).or_insert(0) += 1;
            if let Some(left) = &node.left {
                queue.push_back(left.clone());
            }
            if let Some(right) = &node.right {
                queue.push_back(right.clone());
            }
        }

        let max = *freq.values().max().unwrap_or(&0);
        freq.iter()
            .filter(|(_, &v)| v == max)
            .map(|(&k, _)| k)
            .collect()
    }
}
